neat_ml/tests/test_workflow.py

```diff
@@ -0,0 +1,179 @@
+    ds = {"id": "DS3", "method": "OpenCV", "detection": {"img_dir": str(tmp_path)}}
+    paths = {"proc_dir": tmp_path / "p"}  # det_dir missing
+
+    wf.stage_opencv(ds, paths)
+
+    assert "Detection paths not built (step not selected or misconfig). Skipping." in caplog.text
+
+
+def test_stage_opencv_warns_when_img_dir_missing(caplog: pytest.LogCaptureFixture, tmp_path: Path) -> None:
+    """
+    stage_opencv: if detection.img_dir missing -> warning and return.
+    """
+    caplog.set_level(logging.WARNING)
+    ds = {"id": "DS4", "method": "OpenCV", "detection": {}}
+    paths = {"proc_dir": tmp_path / "p", "det_dir": tmp_path / "d"}
+
+    wf.stage_opencv(ds, paths)
+
+    assert "No 'detection.img_dir' set for dataset 'DS4'. Skipping detection." in caplog.text
+
+
+def test_stage_opencv_skips_if_output_already_exists(
+    caplog: pytest.LogCaptureFixture, 
+    tmp_path: Path, 
+    monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """
+    stage_opencv: if *_bubble_data.pkl exists -> skip and DO NOT call pipeline steps.
+    """
+    caplog.set_level(logging.INFO)
+
+    proc_dir = tmp_path / "proc"
+    det_dir = tmp_path / "det"
+    det_dir.mkdir(parents=True)
+    (det_dir / "anything_bubble_data.pkl").write_text("done")
+
+    def bad(*_: Any, **__: Any) -> None:  # pragma: no cover
+        raise AssertionError("Pipeline function should not be called when outputs exist")
+
+    monkeypatch.setattr(wf, "cv_preprocess", bad)
+    monkeypatch.setattr(wf, "collect_tiff_paths", bad)
+    monkeypatch.setattr(wf, "run_opencv", bad)
+
+    ds = {"id": "DS5", "method": "OpenCV", "detection": {"img_dir": str(tmp_path)}}
+    paths = {"proc_dir": proc_dir, "det_dir": det_dir}
+
+    wf.stage_opencv(ds, paths)
+
+    assert "Detection already exists for DS5. Skipping." in caplog.text
+
+
+def test_stage_opencv_happy_path_calls_pipeline(
+    tmp_path: Path, 
+    monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """
+    stage_opencv: happy path creates directories and calls 
+    preprocess -> collect -> run_opencv.
+    """
+    proc_dir = tmp_path / "proc"
+    det_dir = tmp_path / "det"
+    img_dir = tmp_path / "imgs"
+    img_dir.mkdir()
+
+    calls: list[str] = []
+
+    def fake_cv_preprocess(src: Path, dst: Path) -> None:
+        assert src == img_dir and dst == proc_dir
+        calls.append("preprocess")
+
+    def fake_collect(p: Path) -> list[Path]:
+        assert p == proc_dir
+        calls.append("collect")
+        return [proc_dir / "a.tiff", proc_dir / "b.tiff"]
+
+    def fake_run(
+        df: dict[str, Any], 
+        output_dir: Path,
+        *, 
+        debug: bool = False
+    ) -> None:
+        assert len(df) == 2
+        assert output_dir == det_dir
+        assert debug is True
+        calls.append("run")
+
+    monkeypatch.setattr(wf, "cv_preprocess", fake_cv_preprocess)
+    monkeypatch.setattr(wf, "collect_tiff_paths", fake_collect)
+    monkeypatch.setattr(wf, "run_opencv", fake_run)
```

Comment:
This all looks quite hacky with the monkeypatching--is it really needed? I kind of doubt it -- maybe you should try to describe in clear terms what you're trying to accomplish (this is very hard to read/understand/maintain) so it can be simplified.